### ISSUE TITLE
async/await all the way

### ISSUE THREAD

**Issue by fubar-coder** [fluentmigrator/fluentmigrator#854]
I'm not sure if this project would benefit from async/await support.

I have the vague feeling that it might be better to go async/await, because DB-oriented libraries will go async-first (in the future) and simply use `.Wait()` for the non-async functions. On the other hand, we're usually not blocking the application/server, because the migrations will usually be done beforehand manually (migration tool) or at startup.

Here is what I found out about the ADO.NET providers:

ADO.NET provider | Version | Direct async support
-----------------|-----------|-------------
FirebirdSql.Data.FirebirdClient | 6.0.0-beta1 | No
Npgsql | 4.0.0-preview2 | Yes
System.Data.SqlClient | 4.4.3 | Yes <sup>1</sup>
System.Data.SQLite | 1.0.108 | No
MySql.Data |  8.0.10-rc | No <sup>2</sup>
Oracle.ManagedDataAccess | 12.2.1100 | No
IBM.Data.DB2 | 11.1.3030.4 | No
Sap.Data.Hana.v4.5 | 2.2.36.0 | No <sup>3</sup>

<sup>1</sup> Creates async tasks by wrapping the old-style `Begin*`/`End*` functions
<sup>2</sup> Provides `Begin*`/`End*` functions for some platforms, but no `*Async` functions
<sup>3</sup> Provides `Begin*`/`End*` functions and (partially) `*Async`, but executes the queries synchronously

Every ADO.NET provider gains some kind of async support through `DbCommand`, but this might not be as performant as it could've been.

**Comment by jzabroski** [comment]
I have found a use case for async/await:

Control+C to cancel the migration pipeline.  Without a CancellationToken to pass, there is no good way other than closing the terminal to do this

**Comment by jzabroski** [comment]
Ran into the need for this again.

**Comment by Pretasoc** [comment]
I`d love to see async support and have just started working on it, but as it is a rather large feature, it will take some time until a PR is ready.

**Comment by jzabroski** [comment]
Hi @pretasoc

I'm in the process of an enormous PR right now so it makes sense to coordinate any other large efforts. My PR is to correctly support  .net core 3.1. As part of this I'm also restructuring the unit tests. In addition, there are some basic things I want to fix before adding async support, as making broken things async likely only makes some things worse.

**Comment by Pretasoc** [comment]
Hi @jzabroski,

good to know, that there are other changes ahead. Almost all changes i made so far are currently pushed to the fork at my Github. Basically i am trying to provide an asynchronous interface for each component, that either communicates with the database or uses such a component. My goal is to implement theese interfaces in the existing classes, so we can avoid code duplication.

I know a maybe started a little fast on that PR. But im trying to bring fluent migrator into my company as we are currenly exploring different database migration options for one of out applications, and async support would defenetly be a huge plus.

Until now i primary touched the `IMigrationProcessor` implementations, but there are a lot of implementations, that mostly all need a change to support async.

Currently i could create a small PR, that only mimic async support, but in fact calls executes the synchronous code under an async wrapper.

**Comment by jzabroski** [comment]
> Almost all changes i made so far are currently pushed to the fork at my Github.

I found your branch: https://github.com/Pretasoc/fluentmigrator/tree/feature/AsyncMigrator

I am thankful you put descriptive comments in your commits.  But the real work is likely stuff like:

1. Handling Control+C SIGINT when running FluentMigrator.DotNet.Cli, FluentMigrator.MSBuild, or FluentMigrator.Console from the command line. - This is going to be a new feature and so the framework needs new test cases to cover things like verifying rolling back all transactions, verifying we log to the journal table (VersionInfo table) the progress made so far when using transaction per migration, etc.  All the things you take for granted when you run everything synchronously.
2. What does it mean to execute asynchronously a Migration Constraint before running the migration.  You have to get the result of that check prior to actually running the migration or not.  This could get messy, fast, as the code in this area of the system is a mix-and-match of many different open source contributors over the years, and needs cleaning up.

> I know a maybe started a little fast on that PR.

Well, yes, I like the enthusiasm but understand there are two ways to approach these problems:

1. First analyze that all the database driver dependencies actually support async/await all the way back to 4.6.1.
2. Assume some drivers just might not have support for async/await, especially in 4.6.1.  Propose plan for how to deal with database drivers with no async/await support.

There is the additional stuff I was alluding to, which is that currently we need to clean up "who owns the connection" in FluentMigrator.  Under certain conditions, the last transaction may hang.

> Currently i could create a small PR, that only mimic async support, but in fact calls executes the synchronous code under an async wrapper.

I do not want to maintain hacky async support ("synchronous code under an async wrapper").  We either do it right or we don't do it.  For some context, I probably spend ~300 hours a year on support answering questions and/or investigating bugs and/or reviewing PRs.  The last thing I want to troubleshoot is reports that FluentMigrator deadlocks under some conditions, because neither me nor the customer want to spend time to reproduce those kinds of issues.

The last thing I want to do is build a database migration framework that could deadlock, so the idea of putting an async wrapper on synchronous code is a non-starter.  It's literally just crazy to think anyone would use FluentMigrator if it has a remote chance of deadlock.  Nobody should use such a framework.  The whole reason I started using FluentMigrator in the first place is because I previously used RoundhousE and discovered they had (a) horrible release management practices (b) horrible, terrifying, silent bugs that caused countless problems in my previous organization that were only discovered months after deployments.

"async support" to me does not mean just adding the async keyword to the code base - if people need to integrate async code into FluentMigrator migrations, they can always just ConfigureAwait and return a Task, which gives them more control over underlying thread pool usage than if we added async everywhere.  "async support" means top-down thinking through how everything is going to work and why.  I've started to do this with dependency injection and created an ADR (Architecture Decision Record) so that people can see my thinking on things.  Going forward, all ADRs will be in that folder so that people don't need to search issues for design perspective.

**Comment by jzabroski** [comment]
Actually, can you explain what you mean by calling synchronous code under an async wrapper?  For example, as Mark mentioned in the original post, DbCommand provides some basic async support.  I think maybe I am just confused by what you meant.  Is there an example in your 140 commits so far of what that looks like so we can agree on an approach?

**Comment by Pretasoc** [comment]
I agree, `async` support is hard to get right and easy to get wrong. And i can assure, i dont just want to slap `async` in the code and call it a day. I can understand the code, i wrote so far, looks kind of hacky, buts its in a very early exploring stage, to see, what is actually possible. 

> Actually, can you explain what you mean by calling synchronous code under an async wrapper? For example, as Mark mentioned in the original post, DbCommand provides some basic async support. I think maybe I am just confused by what you meant. Is there an example in your 140 commits so far of what that looks like so we can agree on an approach?

I must admit, i'm a littble confused, where you see 140 commits of me, as i remember only 5 or so on that branch.

With _"calling synchronous code under an async wrapper"_ i meant a `Task` returning method, that runs fully synchronous lie

```
Task Foo()
{
   DoFoo();
   return Task.CompletedTask;
}
```

**Comment by jzabroski** [comment]
This is why I thought you did 140 commits lol

![image](https://user-images.githubusercontent.com/447485/83974583-75174a00-a8bc-11ea-9774-65c283b1dd08.png)

**Comment by Pretasoc** [comment]
So my current idea how to approach this topic is, to add additional interfaces, that provide asynchronous methods for the things, that can be done asynchronously. If a platform does not support async, those interfaces wouldn't be implemented. As i mostly work on SQL-Server i have a very limited view on the other database frameworks, but on SQL server most database calls can be done asynchronously. I didnt plan to perform migrations in paralell, but only use asynchronous database calls, so no thread pool thread is blocked in the meantime.

**Comment by Pretasoc** [comment]
> 
> 
> This is why I thought you did 140 commits lol
> 
> ![image](https://user-images.githubusercontent.com/447485/83974583-75174a00-a8bc-11ea-9774-65c283b1dd08.png)

Ah, yes i branched from develop.

**Comment by Pretasoc** [comment]
So to conclue this: I will draft a proposal how async support could be realized. AS im currently very busy at my daytimer job, I doubt i can show something before next weekend.

**Comment by jzabroski** [comment]
Got it.  I also want to rework develop, as I don't like how previously we were not following a gitflow model and we were doing too much for a major release IMHO.  I want to do smaller major releases going forward.  It's going to be a lot of work to integrate 3.2.7 to 3.3.0, and then 3.3.0 to 4.0.x releases.  (If you want to help with that, that would be awesome, and you may even be able to give me some ideas in the process.)

I am planning to release 3.2.7 today.  Once 3.3.0 is released, I am considering 3.x done barring bugs that format disks :)  And then the goal would be to follow a proper GitFlow branching model.

**Comment by jzabroski** [comment]
> So to conclue this: I will draft a proposal how async support could be realized. AS im currently very busy at my daytimer job, I doubt i can show something before next weekend.

Thanks for your help.  This all sounds good.

**Comment by jzabroski** [comment]
> Task Foo()
> {
>    DoFoo();
>    return Task.CompletedTask;
> }

That is the theoretically correct way, and makes sense to me.  However, it does not answer how to abort in-progress sequence of migrations.  There needs to be a layer inbetween these Task returns that decides whether it should continue or cancel the "interstitial" task upon SIGINT.  This is effectively what xunit.net does for its test runner, as most unit tests don't accept cancellation tokens themselves.

Thanks for making me think about this a bit harder.  If you didn't start this work I probably would've been lazy and not have gotten to it for 3 years lol.

**Comment by jzabroski** [comment]
One last thought. Should the code use partial classes to hold Async interface. This would declutter the code a bit and reduce the weight of lines of code to read. The sync and async interfaces can then share basic reusable snippets like SQL statements to run to check for existence of a sql object in DB

**Comment by Pretasoc** [comment]
So its one week later and i doubt i will complete the proposal tomorrow. Because i want async support with as few changes as possible in existing migrations, there are some difficulties how to do async migrations in a consistent pattern.

**Comment by jzabroski** [comment]
Hi @Pretasoc , awesome.  Also, regarding sharing basic reusable snippets like SQL statements, this need came up recently, as the way FluentMigrator approaches these snippets is not super-great (but not necessarily bad, either).  For example, see: https://github.com/fluentmigrator/fluentmigrator/issues/1263

I believe the reason FluentMigrator uses these out-of-band tests is that most databases require `CREATE TABLE` statements to be the first statement in the transaction.  Even so, what I'd like to see in the future is something like:

```c#
If.Schema("foo").Table("Bar").NotExists()
  .Do((InlineUpMigration d) =>
    d.Table("Bar").InSchema("foo").Column("a").AsInt64().NotNullable());
```
This would have the portion of FluentMigrator that manages the `System.Data` layer in charge of tracking the "decision tree" for whether a migration should be processed.  In theory, I could then unify this with the rather inconsistent `MigrationConstraint` attribute for selectively executing migrations.  Neither IQuerySchema nor MigrationConstraints support a "live" "decision tree".  Both are snapshots at a particular point in time, which I think is rather clunky despite acknowledging the above reasons as to why it was likely coded this way: It's clunky because it limits the usefulness of generating preview migrations, as well as just writing code that works the way you might naturally want to read it working.  I think we can do better.

Anyway, not really related to async/await, but just giving you an idea of some other thoughts I have in mind for simplifying the overall architecture and pros and cons of how we structure files in the solution.

**Comment by snakex64** [comment]
Has there been any new work on this? Pretty much everything is always async everywhere these days, it's a bit weird to start running SQL commands in a synchronous manner

**Comment by PhenX** [comment]
Async is primarily useful for highly concurrent requests scenarios, I'm not sure FluentMigrator is often in that case : migrations are usually ran before anything else, and even if ran while the application is running it may be via a different connection. 

Also, even if not significative, async adds a very slight overhead in terms of memory and times.

**Comment by jzabroski** [comment]
There is really only two reasons async await would be useful:

1. Execute.WithConnection and injecting your own data access layer that in turn uses async methods - if this isn't async, then the control flow has to use RunSynchronously().

2. Being able to cancel Migrations with a terminal interrupt.

I don't currently have the time required for such a big feature given the benefits aren't that big.

If you are interested in working on a PR, we can discuss how it could work before you start. An initial prototype would probably address use case 1 first, as use case 2 is going to break a lot of API surface area.

**Comment by snakex64** [comment]
Hi! Thanks for the super quick feedback guys !

I'm not quite sure to understand where the "only two reasons" comes from. If I can describe our scenarios, maybe that'll be easier to compare with what you have in mind:

1. In a cloud multi tenant scenario you want to be able to run a migration on the fly for a specific tenant only when that tenant upgrades and/or accepts new EULA. You can't have a thread starvation with half your worker thread in blocking state just because 4 out of hundreds of clients are currently doing migrations. DB operations are DB intensive, not CPU intensive. Creating threads is much more expansive than the added cost of async.

2. We've seen many cases where a simple DB timeout wasn't enough and we got stuck for much longer, i.e due to connection issues which are sometimes not intercepted by the generic timeout configuration. Having asynchronous operations with cancellation token helps a lot insure we have control over what ends when

3. During an upgrade sequence there's more than just DB migrations, and we've always shown a progression menu to the end user with an option to cancel the upgrade manually. Linking that button to a simple cancellation token in an async context is very easy

4. Most upgrade scripts are fast but some can take literal hours, a simple global timeout over a migration is not enough to cover the complexity of what is allowed how much time

I hope this helps understand why I was so surprised when I saw there was no async overload to the "MigrateUp()"!

**Comment by eaglestorm** [comment]
Hi,

Where is this at?

If you think this is not necessary I think you miss understand how async await works in dotnet and how dotnet has been built around all IO being async.

From experience migrating an old framework api to dotnet core this is not something that can be ignored.  Under heavy load you will have performance problems as to get the non-async methods in dotnet for IO i think under the hood they do a lot of getawiater().getresult() which is bad and something that should never have been implemented in dotnet core.   

What this means is that by not calling the async methods it introduces a lot of potential issues with port exhaustion and threading etc that can only be avoided by adding in async methods.

**Comment by jzabroski** [comment]
@eaglestorm Happy to discuss.

Are you using a database provider that doesn't use connection pooling? If yes, limiting the size of the pool will prevent spikes. 

Port exhaustion typically occurs, in my experience as a consultant addressing it at multiple clients, when developers explicitly call `new SqlConnection()` without a `using` directive to guarantee the underlying stream is closed and disposed. When it's not implemented this way, the garbage collector has to sweep the object. And it's not uncommon for the connection to end up in Gen1 or Gen2, especially if connections are getting allocated in a loop and "joined" to the same overarching database transaction via Microsoft Distributed Transaction Coordinator.

If you're using Microsoft DI to acquire the connection and dispose of it, then it's guaranteeing Close() will be called when the object falls out of its lifetime scope, and while it's technically possible to configure Microsoft DI to have mismatch of lifetime scopes and thus a memory leak where Close may not get called at an appropriate time, you can guard against this by validating your service collection when you generate the provider. If you're bringing your own  container to Microsoft DI, then it also can't see those registration rules, so your mileage may vary in how useful it is to validate lifetime scopes.

That being said, I was looking into supporting databases where there isn't a standard System.Data interface for .NET, like Databricks or Trino/Athena/Starburst or Cosmos. With those, they don't really have the concept of a connection pool (in the case of Databricks, code can often be running on a Spark cluster although one can use Databricks data governance to query data directly and stream it back) out of the box.

With regards to WHY we say FluentMigrator may not really need support for this, except for the two scenarios I mentioned (which are real pain points). The main reason is that FluentMigrator basically only ever opens one connection at a time and, by design, doesn't execute SQL in parallel. 

That being said, I'm open to trying Copilot and seeing if it can auto-generate a PR (i would love to be able to cancel FluentMigrator.MSBuild.dll tasks). First, we would need a prompt to give it.

**Comment by eaglestorm** [comment]
seriously ?  [asynchronous-programming](https://learn.microsoft.com/en-us/dotnet/csharp/asynchronous-programming/ )

async await isn't so much about parallel programming its about the fact that you are doing a database call that is asynchronous by its nature and if you are not using async await then my understanding is that in an API that request thread is blocked while you do the database call.

Alternatively if you use async await, that thread is non-blocking so its not forced to wait for a response from the database, i.e. the thread can be re-used while its waiting.

Not implementing this significantly effects scalability when using Fluent Migrator from my understanding and despite the effort would consider moving to entity framework code first if this isn't fixed.

**Comment by jzabroski** [comment]
@eaglestorm Ok. I want to be clear you are aware that **if you are facing port exhaustion, it is not because of FluentMigrator.**  We use a single connection. Further, if you're using SQL Server, the connection pool is controlled by the whole connection string effectively, and that includes the pool size parameter. If they're not identical on the same host, it will use a separate pool.

Your understanding on what effects the scalability is not correct. If your team has a port exhaustion issue, I can refer you to performance experts in my network who can help. I have a huge network of database experts across various DBMSes. I can also solve this problem, but I would prefer to keep am arms length relationship between my open source work and my Simmaculate consulting business. However, I also do not want you wasting time thinking FluentMigrator is the root cause if you're facing production issues with port exhaustion. If you are facing port exhaustion, moving to EntityFrameworkCore migrations will likely not solve it.

What async-await helps with is the ergonomics of killing runaway migrations, especially in DEV. Today, if I kick off a migration and have some dumb bug in it, I cannot Control+^Z in the console to quit and gracefully cancel the connection. I have to kill the tab in Windows Terminal and then go in to the admin login of my db server and type `KILL <server_process_id>`.  I want to be clear I would love to add this feature. I personally don't have the time to add it, and the workaround for me is an inconvenience. It's also not as simple to implement as supporting async-await all the way; we would have to have a separate task run in parallel that can call DbCommand.Cancel() when the cancellation token triggers. This is actually non-trivial and **I doubt EntityFrameworkCore does this as a feature, either**.

**Comment by fairking** [comment]
just my few cents, async allows you to gracely/safely terminate the job (and optionally rollback the changes, if supported). I cannot see why not use async all the way with anything which requires to "wait" (db, file, http, etc).

so your code not wasting time while waiting for a response from the db server, but performs other jobs in meantime.
